File: frontend/src/utils/device.js
export function getDeviceInfo(device, browserValue) {
  const value = String(device || "").trim();
  if (!value) {
    return {
      device: "Unknown Device",
      browser: browserValue || "Unknown App",
    };
  }

  const normalized = value.replace(/^Desktop\/Laptop -\s*/i, "");
  if (!normalized.includes("Mozilla/")) {
    return {
      device: normalized,
      browser: browserValue || normalized,
    };
  }

  const browser =
    normalized.includes("Edg/") ? "Edge" :
    normalized.includes("OPR/") || normalized.includes("Opera") ? "Opera" :
    normalized.includes("Chrome/") || normalized.includes("CriOS/") ? "Chrome" :
    normalized.includes("Firefox/") || normalized.includes("FxiOS/") ? "Firefox" :
    normalized.includes("Safari/") ? "Safari" :
    "Unknown Browser";

  const versionMatch =
    normalized.match(/(?:Edg|OPR|Chrome|CriOS|Firefox|FxiOS|Version)\/([\d.]+)/);
  const version = versionMatch ? versionMatch[1].split(".")[0] : "";

  const platform =
    normalized.includes("Windows NT 10.0") ? "Windows 11" :
    normalized.includes("Windows") ? "Windows" :
    normalized.includes("Mac OS X") || normalized.includes("Macintosh") ? "macOS" :
    normalized.includes("Android") ? "Android" :
    normalized.includes("iPhone") ? "iPhone" :
    normalized.includes("iPad") ? "iPad" :
    normalized.includes("Linux") ? "Linux" :
    "Unknown Device";

  return {
    device: `${platform} - ${browser}`,
    browser: browserValue || `${browser}${version ? ` ${version}` : ""}`,
  };
}

export function formatDevice(device, browser) {
  return getDeviceInfo(device, browser).device;
}

export function formatBrowser(device, browser) {
  return getDeviceInfo(device, browser).browser;
}

export function formatSecurityDate(value) {
  if (!value) return "-";

  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return "-";

  return date.toLocaleString("en-GB", {
    day: "2-digit",
    month: "short",
    year: "numeric",
    hour: "2-digit",
    minute: "2-digit",
    hour12: true,
  }).replace(",", "");
}

export function formatDuration(seconds) {
  const totalSeconds = Number(seconds || 0);
  if (totalSeconds <= 0) return "-";

  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const remainingSeconds = totalSeconds % 60;

  if (hours > 0) return `${hours}h ${minutes}m`;
  if (minutes > 0) return `${minutes}m ${remainingSeconds}s`;
  return `${remainingSeconds}s`;
}

export function getStatusClass(status) {
  const normalized = String(status || "").toLowerCase();

  if (normalized === "success" || normalized === "active") {
    return "bg-emerald-500/15 text-emerald-500 border-emerald-500/20";
  }

  if (normalized === "failed") {
    return "bg-rose-500/15 text-rose-500 border-rose-500/20";
  }

  if (normalized.includes("expired")) {
    return "bg-amber-500/15 text-amber-500 border-amber-500/20";
  }

  return "bg-slate-500/15 text-slate-500 border-slate-500/20";
}

File: backend/app/core/device.py
import re


def _browser_version(user_agent: str, marker: str) -> str:
    match = re.search(rf"{re.escape(marker)}/([\d.]+)", user_agent)
    return match.group(1).split(".")[0] if match else ""


def parse_user_agent(user_agent: str | None) -> dict:
    value = (user_agent or "").strip()

    if not value:
        return {
            "device": "Unknown Device",
            "browser": "Unknown App",
            "browser_version": "",
            "browser_display": "Unknown App",
        }

    if "TradeXDesktop" in value:
        return {
            "device": "TradeX Desktop App",
            "browser": "TradeX Desktop App",
            "browser_version": "",
            "browser_display": "TradeX Desktop App",
        }

    if "Edg/" in value:
        browser = "Edge"
        version = _browser_version(value, "Edg")
    elif "OPR/" in value:
        browser = "Opera"
        version = _browser_version(value, "OPR")
    elif "Chrome/" in value or "CriOS/" in value:
        browser = "Chrome"
        version = _browser_version(value, "Chrome") or _browser_version(value, "CriOS")
    elif "Firefox/" in value or "FxiOS/" in value:
        browser = "Firefox"
        version = _browser_version(value, "Firefox") or _browser_version(value, "FxiOS")
    elif "Safari/" in value:
        browser = "Safari"
        version = _browser_version(value, "Version")
    else:
        browser = "Unknown Browser"
        version = ""

    if "Windows NT 10.0" in value:
        platform = "Windows 11" if "Chrome/" in value or "Edg/" in value else "Windows"
    elif "Windows" in value:
        platform = "Windows"
    elif "Mac OS X" in value or "Macintosh" in value:
        platform = "macOS"
    elif "Android" in value:
        platform = "Android"
    elif "iPhone" in value:
        platform = "iPhone"
    elif "iPad" in value:
        platform = "iPad"
    elif "Linux" in value:
        platform = "Linux"
    else:
        platform = "Unknown Device"

    browser_display = f"{browser} {version}" if version else browser

    return {
        "device": f"{platform} - {browser}",
        "browser": browser,
        "browser_version": version,
        "browser_display": browser_display,
    }
